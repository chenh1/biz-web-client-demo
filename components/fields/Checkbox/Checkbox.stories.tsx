import type { CheckboxProps } from './Checkbox';
import React from 'react';
import { Checkbox } from './Checkbox';

type Args = {
  args: CheckboxProps
}

export default {
  component: Checkbox
}

// More on writing stories with args: https://storybook.js.org/docs/react/writing-stories/args
export const CheckboxHyperRed: Args = {
  args: {
    highlightColor: "hyper-red",
    label: "This is the answer",
  },
};
