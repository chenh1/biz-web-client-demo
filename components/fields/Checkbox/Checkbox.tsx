"use client";

import { getTextColorClass, getBorderColorClass } from '../../constants/colors';
import type { CoreColors } from '../../constants/colors';
import React, { useMemo, useState, ChangeEvent } from 'react';
import { CLICK_CLASS, SECTION_VIEW_CLASS } from '../../constants/tagging';

export type CheckboxProps = {
  label: string | number;
  checkboxRef?: React.RefObject<HTMLInputElement>;
  onChange?: (event: ChangeEvent<HTMLInputElement>) => void;
  value?: string | number;
  disabled?: boolean;
  name?: string;
  id?: string;
  highlightColor?: CoreColors;
  gtmClick?: boolean;
  gtmSectionView?: boolean;
  dataGtmId?: string;
}

export const Checkbox: React.FC<CheckboxProps> = ({
  label,
  checkboxRef,
  onChange,
  value,
  disabled,
  name,
  id = "tempId",
  highlightColor = "hyper-red",
  gtmClick,
  gtmSectionView,
  dataGtmId
}) => {
  const [isChecked, setIsChecked] = useState<boolean>(false);

  const computedTextColorClass = useMemo(() => {
    return getTextColorClass(highlightColor);
  }, [highlightColor]);

  const computedBorderColorClass = useMemo(() => {
    return getBorderColorClass(highlightColor);
  }, [highlightColor]);

  return (
    <span className="relative">
      <input
        data-testid="checkbox"
        data-gtm-id={dataGtmId}
        ref={checkboxRef}
        value={value}
        disabled={disabled}
        name={name}
        onChange={e => {
          setIsChecked(e.target.checked);
          onChange && onChange(e);
        }}
        id={id}
        type='checkbox'
        className={`${gtmClick ? CLICK_CLASS : ''} ${gtmSectionView ? SECTION_VIEW_CLASS : ''} h-0 w-0 absolute`}
      />
      <label
        data-gtm-id={dataGtmId}
        htmlFor={id}
        className={`${gtmClick ? CLICK_CLASS : ''} ${gtmSectionView ? SECTION_VIEW_CLASS : ''} font-body relative flex my-[.6em] items-center text-night-black dark:text-max-white transition-colors duration-[250ms] ease-[cubic-bezier(.4,.0,.23,1)]`}
      >
        <span
          className={`pointer-events-none flex justify-center items-center mr-[.9em] w-[1.1em] h-[1.1em] bg-transparent border-2 border-solid rounded-sm cursor-pointer transition-all duration-[250ms] ease-[cubic-bezier(.4,.0,.23,1)] ${isChecked ? `border-[.55em] border-solid ${computedBorderColorClass} animate-shrinkbounce before:content-[""] before:min-w-[7px] before:min-h-[14px] before:absolute before:top-[.7em] before:left-[.4em] before:border-r-[3px] before:border-b-[3px] before:border-solid before:border-transparent before:rotate-45 before:origin-[0% 100%] before:animate-check` : 'border-night-black dark:border-max-white'}`}
        ></span>
        <span className={`pointer-events-none duration-500 ${isChecked ? 'opacity-0' : 'opacity-1'}`}>
          {label}
        </span>
        <ins className={`pointer-events-none absolute block bottom-0 left-[2em] h-0 w-full overflow-hidden no-underline transition-[height] duration-300 ease-[cubic-bezier(.4,.0,.23,1)] ${isChecked ? 'h-full' : ''}`}>
          <i className={`pointer-events-none absolute bottom-0 not-italic ${computedTextColorClass}`}>{label}</i>
        </ins>
      </label>
    </span>
  );
};
