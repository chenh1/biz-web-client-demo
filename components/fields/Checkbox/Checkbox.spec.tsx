import { Checkbox } from "./Checkbox";
import { fireEvent, render } from '@testing-library/react';
import { JSDOM } from "jsdom"
const dom = new JSDOM()
global.document = dom.window.document
global.window = dom.window


it('renders a checkbox input', async () => {
  const { asFragment } = render(
    <Checkbox label="Pizza" />
  );

  expect(asFragment()).toMatchSnapshot();
});

it('shows the custom checkbox when checked', async () => {
  const { getByTestId, asFragment } = render(
    <Checkbox label="Pizza" />
  );

  fireEvent.click(getByTestId("checkbox"))

  expect(asFragment()).toMatchSnapshot();
});
