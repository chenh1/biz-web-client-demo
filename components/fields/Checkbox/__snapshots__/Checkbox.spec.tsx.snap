// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a checkbox input 1`] = `
<DocumentFragment>
  <span
    class="relative"
  >
    <input
      class="  h-0 w-0 absolute"
      data-testid="checkbox"
      id="tempId"
      type="checkbox"
      value=""
    />
    <label
      class="  font-body relative flex my-[.6em] items-center text-night-black dark:text-max-white transition-colors duration-[250ms] ease-[cubic-bezier(.4,.0,.23,1)]"
      for="tempId"
    >
      <span
        class="pointer-events-none flex justify-center items-center mr-[.9em] w-[1.1em] h-[1.1em] bg-transparent border-2 border-solid rounded-sm cursor-pointer transition-all duration-[250ms] ease-[cubic-bezier(.4,.0,.23,1)] border-night-black dark:border-max-white"
      />
      <span
        class="pointer-events-none duration-500 opacity-1"
      >
        Pizza
      </span>
      <ins
        class="pointer-events-none absolute block bottom-0 left-[2em] h-0 w-full overflow-hidden no-underline transition-[height] duration-300 ease-[cubic-bezier(.4,.0,.23,1)] "
      >
        <i
          class="pointer-events-none absolute bottom-0 not-italic text-hyper-red"
        >
          Pizza
        </i>
      </ins>
    </label>
  </span>
</DocumentFragment>
`;

exports[`shows the custom checkbox when checked 1`] = `
<DocumentFragment>
  <span
    class="relative"
  >
    <input
      class="  h-0 w-0 absolute"
      data-testid="checkbox"
      id="tempId"
      type="checkbox"
      value=""
    />
    <label
      class="  font-body relative flex my-[.6em] items-center text-night-black dark:text-max-white transition-colors duration-[250ms] ease-[cubic-bezier(.4,.0,.23,1)]"
      for="tempId"
    >
      <span
        class="pointer-events-none flex justify-center items-center mr-[.9em] w-[1.1em] h-[1.1em] bg-transparent border-2 border-solid rounded-sm cursor-pointer transition-all duration-[250ms] ease-[cubic-bezier(.4,.0,.23,1)] border-[.55em] border-solid border-hyper-red animate-shrinkbounce before:content-[""] before:min-w-[7px] before:min-h-[14px] before:absolute before:top-[.7em] before:left-[.4em] before:border-r-[3px] before:border-b-[3px] before:border-solid before:border-transparent before:rotate-45 before:origin-[0% 100%] before:animate-check"
      />
      <span
        class="pointer-events-none duration-500 opacity-0"
      >
        Pizza
      </span>
      <ins
        class="pointer-events-none absolute block bottom-0 left-[2em] h-0 w-full overflow-hidden no-underline transition-[height] duration-300 ease-[cubic-bezier(.4,.0,.23,1)] h-full"
      >
        <i
          class="pointer-events-none absolute bottom-0 not-italic text-hyper-red"
        >
          Pizza
        </i>
      </ins>
    </label>
  </span>
</DocumentFragment>
`;
